### DISCUSSION TITLE
Dynamically set ImageUploadEditor filename

### DISCUSSION THREAD

**Discussion by flashyjunior** [serenity-is/Serenity#7106] [Help]
Please how can I set ImageUploadEditor filename dynamically in dialog.ts, what I have done is not working

```
constructor() {
    super();

    this.form.RelationChurchMemberId.changeSelect2(e => {

        var memberid = Q.toId(this.form.RelationChurchMemberId.value);
        if (memberid != null) {
            var member = Indotalent.Member.MemberRow.getLookup().itemById[memberid].FullName;
            var photo = Indotalent.Member.MemberRow.getLookup().itemById[memberid].PhotoUrl;
            this.form.RelationName.value = member;
            this.form.PhotoUrl.value.Filename = photo;
        }
    });
}
```

**Comment by flashyjunior** [comment]
Please any help will be greatly appreciated @VictorTomaili

**Reply by LSNicholls** [reply]
Hi,

I haven't done exactly this (made the change in the ts), but I do dynamically set the upload filename by creating it on the SQL side, as one or more columns (the column values are expressions; they're not really stored in the db).  

I use the {4} argument of the FilenameFormat template plus the pipe-delimited folder path options for this.   Here is an example.

` FilenameFormat = "Projects/|ProjectFolder|/|Subfolder|/{4}"`

My requirements are pretty extensive; I can't see for sure from your code but it does look like you could do it, because you are querying columns from the db.  If you state your requirement in words, I'll explain how I think it would work for your case.

Or, you could read these, which is how I got the idea of what is possible.

https://serenity.is/blog/2017/11/02/upload-file-name-formats
https://serenity.is/Docs/tutorials/movies/12-adding-primary-and-gallery-images

**Comment by flashyjunior** [comment]
Thank you @LSNicholls ,
I want to change the image dynamically based on the user selected from the drop down.
When the user selects a particular **Relation's Name**, the system should load automatically into the **ImageUploadEditor** the image stored for that person in the database

![image](https://github.com/serenity-is/Serenity/assets/7760804/3c65d6b4-c77b-4a1c-a8f6-f4bb3e17451a)

**Comment by LSNicholls** [comment]
I see.  Well, using values from the row will not work, **unless** the dropdown value is persisted to the database first.  

So... Here are some questions for you -- assuming you really do want to store this value, which I'm not sure you need to do at all:

1.  Is the Relation's name updateable after a new row is created?  It seems to me that it would not be, because otherwise your filename/path is going to be wrong if the value is changed later.
2.  Assuming not, how would you feel about hiding or disabling Relation Picture for new records, so that they can't upload until they have picked their Relation?  Typically they would do this by hitting the "Apply Changes" button; they wouldn't have to leave the dialog and come back in.
3. If the answer to the last question is "yes that's acceptable" -- and frankly I think it makes sense from a user experience perspective -- you could use the method that I have outlined.  You just need an appropriate column on your row (probably an Expression) to use in your FilenameFormat attribute.

However... I don't really understand why you are using an UploadEditor control here.  You already have the files, you know what they are, and the user isn't going to upload anything.  Why wouldn't you use an img? If you still want to store the source URL for the img in the database row,  questions 1 and 2 above would still apply, but I would get rid of your upload control and embed the image instead. You just need a div or something that you dynamically adjust. (Use StaticTextBlock for this, in other words.)

With this approach, it seems to me that you could do what you originally wanted to do (load the thing in script in the TS)  instead of generating the value as a column in the row, if you want.  The timing wouldn't matter as much, so questions 1 and 2 wouldn't apply.

**Comment by flashyjunior** [comment]
Hi @LSNicholls,
I only want to display the image if it has already been saved, if it hasnt been saved for that relation, then an upload will be done to save the picture for that relation

**Comment by LSNicholls** [comment]
Well, then I think using the ImageUploadEditor with FilenameFormat template should work for you.  (Read the links that I included in my first response, and tell me if you don't understand how this will work.)  

You just need to use this.isNew() in the dialog's afterLoadEntity(), I think, to determine when you need to disable or hide/remove the upload editor control because the relation hasn't been selected yet.  Refer to my questions 1 and 2, in the second comment.

**Comment by flashyjunior** [comment]
Ok, will try that, but how do i use the FilenameFormat in the .ts file to load the image when the Relation is selected because it seems the FilenameFormat property is only at the server side .cs

**Comment by LSNicholls** [comment]
You don't use it in the .ts file.  You don't have to do anything in the .ts file. It's in the CS but it is dynamic, because you reference a column in the row (use the pipe delimiters), and the value in this column is determined by the relative that they chose.  It's a little hard to explain without knowing your data structures, but it is absolutely possible.  

Also, I'm  not clear on whether Abigail is another application user, but that is what it sounds like, from your first explanation.  So, presumably, if Abigail has uploaded a user image already for herself (I have a "My Profile" dialog for stuff like this, not clear how you do it), you can join from your RelativeId to the UserId in the Users table, and pick up this filename value.  

If this filename already exists, because Abigail or an administrator has attached a photo to her User row, did you want the relative to see it?  Or do you want every relative to always upload their own photo of their relative to this row?

If the latter, then I think you need a way to distinguish between different photos of the same relative, for each member who might be related to the same person.

IOW: Let's say your folder (under ~/App_Data/upload/, or whatever your upload path is) is called RelativeImages.  You will need either a subfolder for each user here, so each one who is related to Abigail has their own photo of her, or you need to use a prefix for each person, rather than just Abigail's name, to ensure they stay unique.  For example, if your current User's UserId is 53, that User's photo for Abigail could be RelativeImages/53/AbigailAgyapong.png, or it could be RelativeImages/53-AbigailAgyapong.png.

Does this make sense so far?  If using the separate folders, your FilenameFormat attribute might look like this:

`RelativeImages/|UserId|/|RelativeNameIncludingFileExtension|`

OTOH, if you are using the same photo for all people related to Abigail, and you have that value in PhotoUrl, that's fine too.  Is PhotoUrl pathed?  If so, you'll just want the filename without the path (parse it out if necessary) .  Your FilenameFormat attribute might look like this:

`UserImages/|PhotoUrlFileNameOnlyIncludingExtension|`
 
I am sorry if I am confusing you.  There are a couple of things about your description and the workflow that is supposed to take place that I am having trouble understanding.  I hope my questions in this comment help you see why.

Also, you notice that I say "IncludingFileExtension".  If you are expecting people to upload images BUT you are also pre-selecting what the filename should be, before they upload, I think you are going to have to either limit the allowed Extensions to exactly 1 type (using the ImageExtensions attribute) or you are going to have to post-fix the filename value after they upload, to use the extension that they ended up uploading (you will have it, if you use a separate column to store OriginalName, this is another attribute on the editor).
